pytube/parser.py

```diff
@@ -0,0 +1,83 @@
+import ast
+import json
+import re
+from pytube.exceptions import HTMLParseError
+
+
+def parse_for_object(html, preceding_regex):
+    """Parses input html to find the end of a JavaScript object.
+
+    :param str html:
+        HTML to be parsed for an object.
+    :param str preceding_regex:
+        Regex to find the string preceding the object.
+    :rtype dict:
+    :returns:
+        A dict created from parsing the object.
+    """
+    regex = re.compile(preceding_regex)
+    result = regex.search(html)
+    if not result:
+        raise HTMLParseError(f'No matches for regex {preceding_regex}')
+
+    start_index = result.span()[1]
+    return parse_for_object_from_startpoint(html, start_index)
+
+
+def parse_for_object_from_startpoint(html, start_point):
+    """Parses input html to find the end of a JavaScript object.
+
+    :param str html:
+        HTML to be parsed for an object.
+    :param int start_point:
+        Index of where the object starts.
+    :rtype dict:
+    :returns:
+        A dict created from parsing the object.
+    """
+    html = html[start_point:]
+    if html[0] != '{':
+        raise HTMLParseError('Invalid start point.')
+
+    # First letter MUST be a open brace, so we put that in the stack,
+    # and skip the first character.
+    stack = ['{']
```

Comment:
hmm why not use the regex instead of the while loop here? bit confused as to the purpose / additional value

Reply:
See https://github.com/nficano/pytube/issues/836